homeassistant/components/thethingsnetwork/sensor.py

```diff
@@ -1,165 +1,53 @@
+    """Represents a TTN Home Assistant Sensor."""
 
     @property
-    def native_value(self):
+    def native_value(self) -> float | int | str:
         """Return the state of the entity."""
-        if self._ttn_data_storage.data is not None:
-            try:
-                return self._state[self._value]
-            except KeyError:
-                return None
-        return None
-
-    @property
-    def native_unit_of_measurement(self):
-        """Return the unit this state is expressed in."""
-        return self._unit_of_measurement
-
-    @property
-    def extra_state_attributes(self):
-        """Return the state attributes of the sensor."""
-        if self._ttn_data_storage.data is not None:
-            return {
-                ATTR_DEVICE_ID: self._device_id,
-                ATTR_RAW: self._state["raw"],
-                ATTR_TIME: self._state["time"],
-            }
-
-    async def async_update(self) -> None:
-        """Get the current state."""
-        await self._ttn_data_storage.async_update()
-        self._state = self._ttn_data_storage.data
-
-
-class TtnDataStorage:
-    """Get the latest data from The Things Network Data Storage."""
-
-    def __init__(self, hass, app_id, device_id, access_key, values):
-        """Initialize the data object."""
-        self.data = None
-        self._hass = hass
-        self._app_id = app_id
-        self._device_id = device_id
-        self._values = values
-        self._url = TTN_DATA_STORAGE_URL.format(
-            app_id=app_id, endpoint="api/v2/query", device_id=device_id
-        )
-        self._headers = {ACCEPT: CONTENT_TYPE_JSON, AUTHORIZATION: f"key {access_key}"}
-
-    async def async_update(self):
-        """Get the current state from The Things Network Data Storage."""
-        try:
-            session = async_get_clientsession(self._hass)
-            async with asyncio.timeout(DEFAULT_TIMEOUT):
-                response = await session.get(self._url, headers=self._headers)
-
-        except (TimeoutError, aiohttp.ClientError):
-            _LOGGER.error("Error while accessing: %s", self._url)
-            return None
-
-        status = response.status
-
-        if status == HTTPStatus.NO_CONTENT:
-            _LOGGER.error("The device is not available: %s", self._device_id)
-            return None
-
-        if status == HTTPStatus.UNAUTHORIZED:
-            _LOGGER.error("Not authorized for Application ID: %s", self._app_id)
-            return None
-
-        if status == HTTPStatus.NOT_FOUND:
-            _LOGGER.error("Application ID is not available: %s", self._app_id)
-            return None
-
-        data = await response.json()
-        self.data = data[-1]
-
-        for value in self._values.items():
-            if value[0] not in self.data:
-                _LOGGER.warning("Value not available: %s", value[0])
-
-        return response
+        return cast(float | int | str, self._ttn_value.value)
```

Comment:
Does it work without cast?

Reply:
no, mypy complains without the cast:

`homeassistant/components/thethingsnetwork/sensor.py:53: error: Returning Any from function declared to return "str | int | float | None"  [no-any-return]
Found 1 error in 1 file (checked 6 source files)`

The alternative was the internmediate variable I had here before.

Reply:
Is the library fully typed? Consider adding a py.typed file, this allows mypy to see the types of the library and return a correct value

Reply:
it is not typed yet but it can - I will add types there so it can be nativelly report the right types for the value.

Reply:
I mean, it's not a requirement, but it would make stuff cleaner, and that's always nice :)

Reply:
actually I already had different types in the client library for `TTNSensorValue` vs the generic `TTNBaseValue`. Currently, since we only support Sensors, the base `TTNEntity` is define ttn_value as `TTNSensorValue`.

I am not completely sure why mypy is thing the type of the ttn_value.value is `tuple[<typing special form>, Any]` . Perhaps because the `@propery` decorator. I will investigage it but I am not sure I can avoid the cast X-D

Reply:
Does `TTNSensorValue` inherit `TTNBaseValue`? Then you could overwrite types in the subclass.

But without py.typed, mypy can't see the type of ttn_value.value.

I think it helps to remove the integration from `.strict-typing` until you fixed the `py.typed` thing

Reply:
![image](https://github.com/home-assistant/core/assets/4406403/7bd096cf-37e7-4d74-b3d3-ce4c0782140a)

It should work...

But since it is the only cast in the whole code (and we now is a NOP) it would be a pitty to dissable `.strict-typing` - it would then mean no checks at all instead of a single location IMO. What do you think?

Reply:
I suppose is because I do not export the types from the client library - I will add them there.

Reply:
Afted adding the types to the client library the cast is not longer needed.